Add minimum length optimization to RegexCompiler

stephentoub opened the pull request:
We now do a cursory examination of the RegexNode tree in order to determine the smallest length string that could possibly match.  For example, the regex "(?:123|4567)8" has a minimum length of 4 (if the input were "1238"), and the regex "\d{1-2}-\d{1-2}-\d{2-4}" has a minimum length of 6 (e.g. "1-10-20"). If the input string is shorter than that minimum length, FindFirstChar will now immediately return false, as it can't possibly match, avoiding spending time evaluating a pattern that couldn't possibly match.

Contributes to https://github.com/dotnet/runtime/issues/1349
cc: @danmosemsft, @eerhardt, @ViktorHofer, @pgovind 

As a silly example:
```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Text.RegularExpressions;

public class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssemblies(new[] { typeof(Program).Assembly }).Run(args);

    [Params(RegexOptions.None, RegexOptions.Compiled)]
    public RegexOptions Options { get; set; }

    private string _input = new string('x', 28);
    private Regex _regex;

    [GlobalSetup]
    public void Setup() => _regex = new Regex(@".*hello there how are you today", Options);

    [Benchmark] public bool Test() => _regex.IsMatch(_input);
}
```

| Method |           Toolchain |  Options |        Mean |      Error |     StdDev |      Median | Ratio |
|------- |-------------------- |--------- |------------:|-----------:|-----------:|------------:|------:|
|   Test | \master\corerun.exe |     None | 7,692.43 ns | 193.846 ns | 456.919 ns | 7,525.43 ns | 1.000 |
|   Test |     \pr\corerun.exe |     None |    24.13 ns |   0.104 ns |   0.098 ns |    24.13 ns | 0.003 |
|        |                     |          |             |            |            |             |       |
|   Test | \master\corerun.exe | Compiled | 2,141.20 ns |  12.438 ns |  11.026 ns | 2,141.36 ns |  1.00 |
|   Test |     \pr\corerun.exe | Compiled |    24.38 ns |   0.124 ns |   0.110 ns |    24.37 ns |  0.01 |

stephentoub commented:
(I'm going to tweak this to add it to the interpreter as well.)

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs:
Do we have sufficient test coverage of such degenerate expressions?

stephentoub replied:
I'll add one if we don't.

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs:
```diff
@@ -1424,6 +1484,107 @@ protected void GenerateInitTrackCount()
 
+        /// <summary>Computes a min bound on the required length of any string that could possibly match.</summary>
+        /// <returns>The min computed length.  If the result is 0, there is no minimum we can enforce.</returns>
+        protected int ComputeMinLength(RegexNode node, int depth = 0)
```
Should this be properly on the RegexNode class?

stephentoub replied:
Yes, I have that locally, as part of https://github.com/dotnet/runtime/pull/1624#issuecomment-573379306.

danmosemsft replied:
Also , this might be worth testing through reflection - thoughts? It would be more reassuring than relying on the existing test cases to discover it from outside in . It seems a well defined method to test.

stephentoub replied:
I'd rather not.  It's impact can be pretty easily validated via regexes.  Why is that concerning?

danmosemsft replied:
Fair enough. I guess I'm just worried about yet more branches to test - might we have test regexes that will no longer exercise what they did before, as they now fall out of this fast path?

stephentoub replied:
> as they now fall out of this fast path?

It's possible, but for that to happen the inputs would never have been able to match; they'd get part of the way through the pattern, and then there'd be nothing left.  Regardless of this change, we should ensure we're testing our regexes against inputs that are too short, exactly right, and longer than needed, and this change will just make some of the "too short" ones execute faster.

danmosemsft commented:
BTW, what does RegexFCD mean? First Character Detector?

stephentoub commented:
> BTW, what does RegexFCD mean? First Character Detector?

I honestly have no idea about the 'D'. :)  The "FC" does mean "First Characters", at least based on context.  I've gone through and cleaned up a decent amount of such stuff previously, but there's still more to be done. :)

stephentoub committed:
Add minimum length optimization to regex

We now do a cursory examination of the RegexNode tree in order to determine the smallest length string that could possibly match.  For example, the regex "(?:123|4567)8" has a minimum length of 4 (if the input were "1238"), and the regex "\d{1-2}-\d{1-2}-\d{2-4}" has a minimum length of 6 (e.g. "1-10-20"). If the input string is shorter than that minimum length, FindFirstChar will now immediately return false, as it can't possibly match, avoiding spending time evaluating a pattern that couldn't possibly match.
